libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,1018 @@
+    *
+    * @brief Defines a deleted index on an EOSIO Key Value Table
+    * @details Due to the way indexes are named, when deleting an index a "placeholder" index needs to be created instead.
+    * A null_index should be created in this case. If using DEFINE_TABLE, just passing in nullptr will handle this.
+    */
+   class null_index{};
+
+   /**
+    * @ingroup keyvalue
+    * Puts a value into the table. If the value already exists, it updates the existing entry.
+    * The key is determined from the defined primary index.
+    * If the put attempts to store over an existing secondary index, the transaction will be aborted.
+    *
+    * @param value - The entry to be stored in the table.
+    */
+   void put(const T& value) {
+      uint32_t value_size;
+      T old_value;
+
+      auto primary_key = primary_index->get_key(value);
+      auto tbl_key = table_key(make_prefix(table_name, primary_index->index_name), primary_key);
+      auto primary_key_found = internal_use_do_not_use::kv_get(db_name, contract_name.value, tbl_key.data(), tbl_key.size(), value_size);
+
+      if (primary_key_found) {
+         void* buffer = value_size > detail::max_stack_buffer_size ? malloc(value_size) : alloca(value_size);
+         auto copy_size = internal_use_do_not_use::kv_get_data(db_name, 0, (char*)buffer, value_size);
+
+         deserialize(old_value, buffer, copy_size);
+
+         if (value_size > detail::max_stack_buffer_size) {
+            free(buffer);
+         }
+      }
+
+      for (const auto& idx : secondary_indices) {
+         auto sec_tbl_key = table_key(make_prefix(table_name, idx->index_name), idx->get_key(value));
+         auto sec_found = internal_use_do_not_use::kv_get(db_name, contract_name.value, sec_tbl_key.data(), sec_tbl_key.size(), value_size);
+
+         if (!primary_key_found) {
+            eosio::check(!sec_found, "Attempted to store an existing secondary index.");
+         } else if (sec_found) {
+            void* buffer = value_size > detail::max_stack_buffer_size ? malloc(value_size) : alloca(value_size);
+            auto copy_size = internal_use_do_not_use::kv_get_data(db_name, 0, (char*)buffer, value_size);
+
+            eosio::check(copy_size == tbl_key.size(), "Attempted to update an existing secondary index.");
+            auto res = memcmp(buffer, tbl_key.data(), copy_size);
+            eosio::check(res == 0, "Attempted to update an existing secondary index.");
+
+            if (copy_size > detail::max_stack_buffer_size) {
+               free(buffer);
+            }
+         }
+
+         if (primary_key_found) {
+            auto old_sec_key = table_key(make_prefix(table_name, idx->index_name), idx->get_key(old_value));
+            internal_use_do_not_use::kv_erase(db_name, contract_name.value, old_sec_key.data(), old_sec_key.size());
+         }
+
+         internal_use_do_not_use::kv_set(db_name, contract_name.value, sec_tbl_key.data(), sec_tbl_key.size(), tbl_key.data(), tbl_key.size());
```

Comment:
If the secondary key is not modified (probably the common case), this causes an unnecessary erase+set.